### DISCUSSION TITLE
Text width and height based on number of characters / lines displayed

### DISCUSSION THREAD

**Discussion by vikigenius** [hecrj/iced#1201] [Q&A]
So I am trying to develop a card like item that contains long texts, that can be expanded on a button click.

Currently I limit the size of the card by fixing the number of characters I display. (Can just take a slice). Then on button click I just change the text.

```rs
        let ctext = if self.expanded { self.text.clone() } else { self.text[..MAX_RENDER_CHARS].to_owned() };
```

But what if I want to fix the number of lines that I want to display? Say I want to display the first 4 lines? 

Equivalently, how do I determine the number of characters that will fit in a single line given width and height of the text element? Is this possible?

**Comment by sunny-chung** [comment]
To make advanced text-related widgets and complex UI, it is required to have functions able to measure text metrics such as character width and height of individual characters, especially when CJK characters and emoji sequences are involved. Unfortunately, it is not available now. 

If it is complicated to implement, please at least expose some low-level API calls so that we could try to implement ourselves.

**Reply by EmmanuelDodoo** [reply]
You could create a `iced::advanced::text::paragraph::Plain` and get the overall minimum bounds of the text. I don't think `cosmic-text` under the hood provides metrics for individual characters.

**Reply by sunny-chung** [reply]
Thanks! I gave it a try and the API is adequate for my usage. The measurement result is very accurate.
